ROX-17224: use correct Prometheus client CA signer

stehessel opened the pull request:
## Description

I noticed that I used the wrong signer in #6796 and found an interesting behavior of the TLS config holder. If no client CAs have been found, it defaults to `ClientAuth = NoCertGiven`, which is why the wrong signer had no effect. I changed this behavior for the secure metrics endpoint to `RequireAndVerifyClientCert`. I can't really say if the existing behavior makes sense for Central mTLS, so I left it as is - only made it explicit.

I also added a test to verify that the client auth is not changed if no suitable client CA has been found.

Excerpt from the CA and client cert to make sure it's correct this time:

Client CA:
```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 8831346098925119782 (0x7a8f3e7c8af6a926)
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN=openshift-kube-controller-manager-operator_csr-signer-signer@1688457768
        Validity
            Not Before: Jul  4 08:04:45 2023 GMT
            Not After : Aug  3 08:04:46 2023 GMT
        Subject: CN=kube-csr-signer_@1688457885
...
```

client cert:
```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            63:58:e3:73:2d:9b:c4:f1:ef:d3:cd:cd:9d:35:99:e9
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN=kube-csr-signer_@1688457885
        Validity
            Not Before: Jul  4 11:19:38 2023 GMT
            Not After : Aug  3 08:04:46 2023 GMT
        Subject: CN=system:serviceaccount:openshift-monitoring:prometheus-k8s
```

## Checklist
- [ ] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

stehessel commented on pkg/metrics/tls.go:
```diff
@@ -181,10 +182,13 @@ func (t *tlsConfigurerImpl) updateClientCA(cm *v1.ConfigMap) {
 		var signerCAs []*x509.Certificate
 		for _, c := range certs {
-			if c.Issuer.CommonName == signerName {
+			if strings.HasPrefix(c.Subject.CommonName, signerName) {
```
I also wonder how stable that subject name is. Thoughts on just increasing the client CA scope to all certificates in `extension-apiserver-authentication`?

dhaus67 replied:
Hmm, I think the argument could be made for both sides regarding whether to consider this as stable.

It could be sufficient for now to verify the certificate itself against the cert pool created out of the CA of `extension-apiserver-authentication` (incl. the verification for the key usage), looking at the code of Kubernetes we are essentially following [the same principles there](https://github.com/kubernetes/apiserver/blob/master/pkg/server/dynamiccertificates/tlsconfig.go#L170-L187) (and looking at [the documentation for client auth within the api server](https://github.com/kubernetes-sigs/apiserver-builder-alpha/blob/master/docs/concepts/auth.md#client-certificate-authentication) that seems to be also somewhat in line).

Within the config map, will all certificates signed there be issued by the `kube-csr-signer`? If that's the case, then the point of doing the extra verification here is moot, since virtually all certificates contained within the map could be used to authZ against this endpoint (even if we may not intend them to).

stehessel replied:
> Within the config map, will all certificates signed there be issued by the kube-csr-signer?

No, there are different issuers, for example `admin-kubeconfig-signer` and `kube-control-plane-signer`.

```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 7874177071138344378 (0x6d46b253b7ba0dba)
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: OU=openshift, CN=kube-control-plane-signer
        Validity
            Not Before: Jul  3 12:50:05 2023 GMT
            Not After : Jul  2 12:50:05 2024 GMT
        Subject: OU=openshift, CN=kube-control-plane-signer
```

stehessel replied:
But it does sound like the expected behavior is to simply verify against the whole cert pool, even if that extends the allowed certificates. So I would probably opt to remove this check again and load all CA certs.

stehessel commented:
Current dependencies on/for this PR:
* master
  * **PR #6846** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6846" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈
    * **PR #6858** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6858" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/6846?utm_source=stack-comment).

dhaus67 commented on central/tlsconfig/manager_impl.go:
```diff
@@ -149,7 +149,7 @@ func (m *managerImpl) TLSConfigurer(opts Options) (verifier.TLSConfigurer, error
 		rootCfg.ClientAuth = tls.RequireAndVerifyClientCert
 	}
-	configurer := certwatch.NewTLSConfigHolder(rootCfg)
+	configurer := certwatch.NewTLSConfigHolder(rootCfg, tls.NoClientCert)
```
I think one thing that _may_ be confusing is that this here AFAICT isn't necessarily the TLS config for sensor connections, but rather the one based on [the endpoint config file](https://github.com/stackrox/stackrox/blob/master/image/templates/helm/stackrox-central/config/central/endpoints.yaml.default) contents and by default the client auth is set to `RequireAndVerifyClientCert` which makes sense due to mTLS for central <-> sensor as well as, if configured, user PKI auth provider.

I'd say the reason why this is causing trouble now is because you have a race-condition in the first call to `UpdateTLSConfig`: you call it twice, either on updates to the server certificates or on updates to the contents of the config map containing the CA. Assumption would be that reading the server certificates within the directories will be faster than obtaining the CA and its contents from Kubernetes, hence the first call to `UpdateTLSConfig` will be when the server certificates are received (and no client CAs are set), thus the `ClientAuth` option is unset / set to `NoClientCert`.

While the behavior isn't what one would expect, it's somewhat understandable why it has been like this before: the mount is either there, or it's not.

Looking at it back now, in your case we are essentially lazy loading the TLS config at the moment, and the initial serve from the metrics endpoint may or may not contain all required certificates (i.e. the `NewTLSConfigurerFromEnv` call). Ideally, we'd have a "run once" watch for changes to be sure we at least attempted _once_ to fetch required certificates before returning the config in this case.

stehessel replied:
The start up is not the only case where the fallback matters. It also comes into play when the config map is deleted / changed such that no certs are found.

> Ideally, we'd have a "run once" watch for changes to be sure we at least attempted once to fetch required certificates before returning the config in this case.

What do we do then if the initial fetch fails? Tbh I don't see an issue with the lazy loading. The question is if incoming requests should be rejected (proposed behavior), or allowed (current behavior), if the certs are either not there or incomplete.

stehessel replied:
> I think one thing that may be confusing is that this here AFAICT isn't necessarily the TLS config for sensor connections, but rather the one based on [the endpoint config file](https://github.com/stackrox/stackrox/blob/master/image/templates/helm/stackrox-central/config/central/endpoints.yaml.default) contents and by default the client auth is set to RequireAndVerifyClientCert which makes sense due to mTLS for central <-> sensor as well as, if configured, user PKI auth provider.

What I find unintuitive is that a TLS config holder is modifying static parts of the TLS config. I think it would make more sense to simply leave `ClientAuth` as it was set in the TLS root config. On the other hand changing this behavior now could cause unintended problems, so I would leave it as is.

dhaus67 replied:
>What do we do then if the initial fetch fails? Tbh I don't see an issue with the lazy loading.

My point is that lazy loading here might be something we have to do due to the nature of reading asynchronously the CA from the config map, however I'd expect that we reject request until we can be sure that we've read the config map successfully _at least_ once and use the _expected_ TLS config. Right now, essentially you won't know which TLS config is being used after serving the first requests, and I don't think we should have this ambiguity here.

>What I find unintuitive is that a TLS config holder is modifying static parts of the TLS config. I think it would make more sense to simply leave ClientAuth as it was set in the TLS root config. On the other hand changing this behavior now could cause unintended problems, so I would leave it as is.

Yeah agreed that part definitely looks weird, but agreed we're better off to not change it for now.

dhaus67 commented on pkg/mtls/certwatch/tls_config_holder.go:
```diff
@@ -24,9 +25,10 @@ type TLSConfigHolder struct {
 
 // NewTLSConfigHolder instantiates a new instance of TLSConfigHolder
-func NewTLSConfigHolder(rootCfg *tls.Config) *TLSConfigHolder {
+func NewTLSConfigHolder(rootCfg *tls.Config, fallbackClientAuthNoCAs tls.ClientAuthType) *TLSConfigHolder {
```
If kept as-is, we should rename this to `defaultClientAuth` or `fallbackClientAuth` with a clarification in the comment how the interaction is currently working.

stehessel commented:
After discussion with @dhaus67 I added two changes:
* `ConfigMapWatcher.init` now attempts to get the config map. It's executed on the main loop and is therefore always executed before the server is started. If the get fails, we log the error and continue.
* Added timeouts to k8s calls in `ConfigMapWatcher` to prevent blocking the main loop.

stehessel commented:
/retest
